File: go/triangle/triangle.go
// Package triangle implements triangle exercism solution from the go track
package triangle

import (
	"math"
	"sort"
)

type Kind string

const (
	NaT Kind = "not a triangle"
	Equ Kind = "equilateral"
	Iso Kind = "isosceles"
	Sca Kind = "scalene"
)

// KindFromSides takes the triangles sides and returns it's kind
func KindFromSides(a, b, c float64) Kind {
	if !isValidTriangle(a, b, c) {
		return NaT
	}
	if (a == b) && (b == c) {
		return Equ
	}
	if (a == b) || (b == c) || (a == c) {
		return Iso
	}
	return Sca
}

func isValidTriangle(a, b, c float64) bool {
	sides := []float64{a, b, c}
	sort.Float64s(sides)
	for _, side := range sides {
		if (side <= 0) || (side == math.Inf(1)) {
			return false
		}
	}
	return sides[0]+sides[1] >= sides[2]
}

File: python/triangle/triangle.py
class TriangleError(Exception):
    pass


class Triangle(object):
    def __init__(self, side_a, side_b, side_c):
        [self.a, self.b, self.c] = sorted([side_a, side_b, side_c])

    @property
    def valid(self):
        if self.a <= 0 or self.b <= 0 or self.c <= 0:
            return False
        if self.a + self.b <= self.c:
            return False
        return True

    @property
    def equilateral(self):
        if not self.valid:
            return False
        return self.a == self.b == self.c

    @property
    def isosceles(self):
        if not self.valid:
            return False
        return self.a == self.b or self.b == self.c

    @property
    def scalene(self):
        if not self.valid:
            return False
        return [self.a, self.b, self.c] == list(set([self.a, self.b, self.c]))


def equilateral(sides):
    triangle = Triangle(*sides)
    return triangle.equilateral


def isosceles(sides):
    triangle = Triangle(*sides)
    return triangle.isosceles


def scalene(sides):
    triangle = Triangle(*sides)
    return triangle.scalene


if __name__ == "__main__":
    pass
